code/game/centcomm_orders/centcomm_orders.dm

```diff
@@ -0,0 +1,386 @@
+
+		if (cargo_contribution > 0 && acct_by_string != "Cargo")//cargo gets some extra coin from every order shipped
+			var/datum/money_account/cargo_acct = department_accounts["Cargo"]
+			cargo_acct.charge(round(-worth * cargo_contribution),null,"Contribution for partial order #[id]",dest_name = name)
+		toPay = 0
+
+// Same as normal, but will take every last bit of what you provided.
+/datum/centcomm_order/per_unit/CheckShuttleObject(var/obj/O, var/in_crate, var/preserve = FALSE)
+	if(must_be_in_crate && !in_crate)
+		return 0
+	if(!O)
+		return 0
+	if(O.type in requested)
+		var/amount = 1
+		if(istype(O, /obj/item/stack))
+			var/obj/item/stack/S = O
+			amount = S.amount
+		if(!(O.type in left_to_check))
+			left_to_check[O.type]=0
+		if (!ExtraChecks(O))
+			return 0
+		left_to_check[O.type] += amount
+		if (!preserve)
+			qdel(O)
+		return 1
+	return 0
+
+/datum/centcomm_order/per_unit/CheckFulfilled()
+	toPay = 0
+	for(var/typepath in left_to_check)
+		var/worth_per_unit = unit_prices[typepath]
+		var/amount         = left_to_check[typepath]
+		toPay += amount * worth_per_unit
+		if(requested[typepath] != INFINITY)
+			requested[typepath] = max(0,requested[typepath] - left_to_check[typepath])
+		if(!(typepath in fulfilled))
+			fulfilled[typepath] = 0
+		fulfilled[typepath] += left_to_check[typepath]
+		left_to_check[typepath] = 0
+	. = ..()
+	Pay(.)
+
+///////////////////////////////
+
+/proc/create_random_orders(var/num_orders)//This one is used at roundstart to add a couple random orders immediately
+	var/list/choices = get_all_orders()
+	for(var/i = 1 to num_orders)
+		var/choice = pick_n_take(choices)
+		var/datum/centcomm_order/new_order = new choice
+		SSsupply_shuttle.add_centcomm_order(new_order)
+
+/proc/get_all_orders()
+	var/list/orders = list()
+	orders.Add(subtypesof(/datum/centcomm_order/per_unit/department/cargo))
+	orders.Add(subtypesof(/datum/centcomm_order/department/science))
+	orders.Add(subtypesof(/datum/centcomm_order/department/medical))
+	orders.Add(subtypesof(/datum/centcomm_order/department/engineering))
+	orders.Add(subtypesof(/datum/centcomm_order/department/civilian))
+	orders.Add(subtypesof(/datum/centcomm_order/per_unit/department/civilian))
+	return orders
+
+
+///////////////////////////////////////
+
+/proc/create_weighted_order()
+	var/list/active_with_role = get_dept_pop()
+
+	var/list/department_weights = list(
+		"Cargo" = 3,
+		"Service" = 5,
+		"Medical" = 5,
+		"Science" = 5,
+		"Engineering" = 5,
+		)
+
+	for(var/dept in department_weights)
+		if(active_with_role[dept] < 1)
+			department_weights[dept] = 1//departments with no employees are very unlikely to receive an order
+
+	var/list/order_exists = list(
+		"Cargo" = 0,
+		"Service" = 0,
+		"Medical" = 0,
+		"Science" = 0,
+		"Engineering" = 0,
+		)
+
+	for(var/datum/centcomm_order/O in SSsupply_shuttle.centcomm_orders)
+		if (istype(O, /datum/centcomm_order/per_unit/department/cargo))
```

Comment:
I would personally do away with the type checks, and do something like
`order_exists[O.dept] += 1`, and dept being a var like "Science", "Medical"... With eventually something to check if the order is not a department-based order.

Reply:
thankfully I can piggyback on the orders acct_by_name var to do that. Just gotta make sure that it's part of the order_exists list since some like the plasma order aren't dept focused.